Add basic StatefulSet tests

TehTotalPwnage opened the pull request:
This PR adds very basic tests to make sure that Longhorn properly integrates with StatefulSets in Kubernetes. The tests make sure that Longhorn volumes used in StatefulSets:
- Mount properly and scale to the correct amount
- Can be scaled up to provision more volumes
- Can be properly backed up and can run recurring backups through Longhorn

This PR implements the first three tests requested in #55. However, the fourth point (restoring into a PVC to be used in a StatefulSet) still needs to be resolved in a separate commit.

yasker commented on manager/integration/tests/common.py:
```diff
@@ -52,19 +52,25 @@
 
 
-def get_core_api_client():
+def load_config():
```
`load_k8s_config`

yasker commented on manager/integration/tests/common.py:
```diff
@@ -383,6 +389,106 @@ def finalizer():
+        api = get_core_api_client()
+        client = get_longhorn_api_client()
+        for i in range(statefulset_manifest['spec']['replicas']):
+            pvc_name = (statefulset_manifest['spec']['volumeClaimTemplates'][0]
```
Is there another way to associate PVC with the stateful set? It's a bit weird if we can only associate them using combined volume name. Is there any tag or fields in the PVC or statefulset we can associate them directly?

yasker replied:
At least seems you already have `get_statefulset_volume_names` below. You can move it up.

yasker commented on manager/integration/tests/test_statefulset.py:
```diff
@@ -0,0 +1,311 @@
+        if s_set.status.ready_replicas == (statefulset_manifest['spec']
+           ['replicas']):
+            break
+
```
Move the blank line one line below. It's a bit misleading here.

yasker commented on manager/integration/tests/test_statefulset.py:
```diff
@@ -0,0 +1,311 @@
+        assert new_b["volumeCreated"] == b["volumeCreated"]
+
+
+def test_statefulset_recurring_backup(clients, core_api, storage_class,  # NOQA
```
tag it with `@pytest.mark.recurring_job` so we can skip it by default. This will take a longer time than other cases.

yasker commented on manager/integration/tests/test_statefulset.py:
```diff
@@ -0,0 +1,311 @@
+    create_storage_class(storage_class)
+    create_and_wait_statefulset(statefulset)
+
+    # backup every two minutes
```
It's every minute you set here.

yasker commented on manager/integration/tests/test_statefulset.py:
```diff
@@ -0,0 +1,311 @@
+    for _, client in clients.iteritems():
+        break
+
+    statefulset_name = 'statefulset-backup-test'
```
looks like this block of code can be a function.

yasker replied:
I think you missed this one?

yasker commented:
BTW, the killing of the pod in stateful set can only be tested with CSI. Our Flexvolume plugin doesn't support it.

TehTotalPwnage commented:
Fixes:
- Marked recurring job test so that it is automatically skipped
- Fixed comment regarding recurring job spec
- Added `get_single_longhorn_client` to avoid repeating the same for loop to get a single Longhorn API client instance. I didn't touch the actual fixture seeing as how it would likely break the other tests.
- `get_statefulset_pod_info` replaces `get_statefulset_volume_names` as it only needs to generate the Pod names, then use the API Pod instances to access PVC and PV information.
- Worked around issue regarding line breaks for the dictionary in the if statement
- Renamed `load_config` to `load_k8s_config`
- Added Pod killing test for CSI

yasker commented on manager/integration/tests/test_statefulset.py:
```diff
@@ -0,0 +1,325 @@
+
+
+def test_statefulset_scaling(clients, core_api, storage_class, statefulset):  # NOQA
+    """
```
Empty comment.

TehTotalPwnage commented:
Fixes:
- Removed `get_single_longhorn_client` due to misunderstood comment. Single Longhorn client fixture can be dealt with in another PR.
- Fixed empty comment for `test_statefulset_scaling`.
- Converted manifest modifying block at the beginning of each test to a dedicated function.

yasker commented:
Merged, thanks.
